R/trace.R

```diff
@@ -295,20 +301,52 @@ jit_trace_module <- function(mod, ..., strict = TRUE) {
     if (!is.list(inp)) {
       inp <- list(inp)
     }
+    if (name == "forward") {
+      tr_fn <- make_traceable_fn(mod[[name]])
+      mod$train()
+      ptr_train <- cpp_trace_function(
+        fn = tr_fn,
+        inputs = inp,
+        compilation_unit = .compilation_unit,
+        strict = strict,
+        module = module$..ptr..(),
+        name = "trainforward",
+        should_mangle = TRUE,
+        manage_memory = FALSE
+      )
+      mod$eval()
```

Comment:
I wonder if we should add an argument to `jit_trace` that would keep the old behavior. My two concerns are:

1. Tracing runs the network twice, which could be problematic for some users.
2. Duplicates the size of the graph, which might be undesidered. Maybe a user wants to just trace the forward method in eval mode to export for deployment.

Reply:
Ok, I changed my mind again, I think the default TRUE is fine, but I am happy to change it as well.
In which cases do you think running the network twice is problematic?

Reply:
Added a `respect_mode` argument that triggers the double/single tracing